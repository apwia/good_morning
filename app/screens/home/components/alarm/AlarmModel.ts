import React from "react";

type AlarmModel = {
	updateAction: any,
	toggleActiveAction: any,
	showTimePick: boolean,
	setShowTimePick: React.Dispatch<React.SetStateAction<boolean>>
	isActive: boolean, 
	displayedTime: string,
	removeAlarmAlert: any,
}

export default AlarmModel;