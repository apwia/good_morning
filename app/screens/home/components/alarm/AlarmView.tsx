import React from 'react';

import { Animated, Switch, Image, Text, TouchableOpacity, View } from 'react-native';

import { GestureHandlerRootView, Swipeable } from 'react-native-gesture-handler';

import Styles from './AlarmStyles';
import useAlarmViewModel from './AlarmViewModel';
import Alarm from '../../../../model/Alarm';
import TimePickerView from '../time_picker/TimePickerView';

type Props = {alarm: Alarm, setShouldRefreshAlarms: React.Dispatch<React.SetStateAction<boolean>>};

const AlarmView = (props: Props) => {
  
  const {updateAction, toggleActiveAction, showTimePick, setShowTimePick, isActive, displayedTime, removeAlarmAlert} 
    = useAlarmViewModel(props.alarm, props.setShouldRefreshAlarms);

  const swipeAction = (_progress: Animated.AnimatedInterpolation<number>, dragX: Animated.AnimatedInterpolation<number>) => {
    
    const scale = dragX.interpolate({
      inputRange: [0, 70],
      outputRange: [0, 1],
      extrapolate: 'clamp',
    });

    return (

      <Animated.View style={[{transform: [{scale: scale}]}, Styles.deleteButtonView, ]}>

        <TouchableOpacity onPress={removeAlarmAlert}>
          <Image 
            style={Styles.deleteIconButton}
            source={require("../../../../assets/images/TrashBin.png")}/>
        </TouchableOpacity>
      </Animated.View>
    );

  };

  return (
    <GestureHandlerRootView>
      <Swipeable
          friction={5}
          renderLeftActions={swipeAction}
          containerStyle={Styles.swipeableContainer}>

        <View style={Styles.alarmView}>

          <View style={Styles.alarmLeftInnerView}>

            <Text onPress={() => {setShowTimePick(true)}} style={Styles.time}>
              {displayedTime}
            </Text>

            <Text style={Styles.periodicity}>{props.alarm.periodicity}</Text>
            
            {showTimePick && 
              <TimePickerView 
                alarm={props.alarm} 
                action={updateAction} 
                whenFinish={()=>{setShowTimePick(false)}} />}

          </View>

          <View style={Styles.alarmRightInnerView}>
            <Switch onValueChange={() => {toggleActiveAction()}} value={isActive} testID='switch'/>
          </View>

        </View>
      </Swipeable>
    </GestureHandlerRootView>
  );
};

export default AlarmView;