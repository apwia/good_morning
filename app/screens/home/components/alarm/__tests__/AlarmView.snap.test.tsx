import 'react-native';

import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react-native';

import AlarmView from '../AlarmView';

// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer';
import AlarmModel from '../AlarmModel';

jest.mock('../AlarmViewModel', () => {
  const fnAlarmViewModel = jest.fn();
  fnAlarmViewModel.mockImplementation(
    () => ({
      updateAction: jest.fn(),
      toggleActiveAction: jest.fn(),
      showTimePick: true,
      setShowTimePick: jest.fn(),
      isActive: true, 
      displayedTime: '12:00',
      removeAlarmAlert: jest.fn(),
    } as AlarmModel),
        
  );

  return fnAlarmViewModel;

});

import useAlarmViewModel from '../AlarmViewModel';

it('should renders correctly', () => {

  const alarm = {	id: 0, time: new Date(1687817628067), active: true, periodicity: 'testing'};

  const props = {alarm: alarm, setShouldRefreshAlarms: jest.fn()}

  const {toJSON} = renderer.create(<AlarmView {...props}/>);
  expect(toJSON()).toMatchSnapshot();

});


it('should render datePick clicking on the text', () => {

  const setShowTimePickFn = jest.fn();

  // @ts-ignore
  useAlarmViewModel.mockImplementation(
    () => ({
      updateAction: jest.fn(),
      toggleActiveAction: jest.fn(),
      showTimePick: true,
      setShowTimePick: setShowTimePickFn,
      isActive: true, 
      displayedTime: '12:00',
      removeAlarmAlert: jest.fn(),
    } as AlarmModel),
  );

  const alarm = {	id: 0, time: new Date(1), active: true, periodicity: 'testing'};

  render(<AlarmView alarm={alarm} setShouldRefreshAlarms={jest.fn()} />);

  fireEvent.press(screen.getByText('12:00'));
  
  expect(setShowTimePickFn).toBeCalledWith(true);

});

it('should call update action after clickin on switch button', () => {

  const toggleActiveActionFn = jest.fn();

  // @ts-ignore
  useAlarmViewModel.mockImplementation(
    () => ({
      updateAction: jest.fn(),
      toggleActiveAction: toggleActiveActionFn,
      showTimePick: true,
      setShowTimePick: jest.fn(),
      isActive: true, 
      displayedTime: '12:00',
      removeAlarmAlert: jest.fn(),
    } as AlarmModel),
  );

  const alarm = {	id: 0, time: new Date(1), active: true, periodicity: 'testing'};

  render(<AlarmView alarm={alarm} setShouldRefreshAlarms={jest.fn()} />);

  const switchComponent = screen.getByTestId('switch');

  fireEvent(switchComponent, 'onValueChange');
  
  expect(toggleActiveActionFn).toBeCalled();

});