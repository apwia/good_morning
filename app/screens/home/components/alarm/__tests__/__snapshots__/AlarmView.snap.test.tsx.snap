// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should renders correctly 1`] = `
<View>
  <View
    collapsable={false}
    forwardedRef={[Function]}
    handlerTag={1}
    handlerType="PanGestureHandler"
    onGestureHandlerEvent={[Function]}
    onGestureHandlerStateChange={[Function]}
    onLayout={[Function]}
    style={
      {
        "flex": 1,
        "overflow": "hidden",
      }
    }
  >
    <View
      collapsable={false}
      style={
        {
          "bottom": 0,
          "flexDirection": "row",
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
          "transform": [
            {
              "translateX": -10000,
            },
          ],
        }
      }
    >
      <View
        collapsable={false}
        style={
          {
            "alignItems": "center",
            "flexDirection": "row-reverse",
            "transform": [
              {
                "scale": 0,
              },
            ],
          }
        }
      >
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "opacity": 1,
            }
          }
        >
          <Image
            source={
              {
                "testUri": "../../../app/assets/images/TrashBin.png",
              }
            }
            style={
              {
                "height": 50,
                "width": 50,
              }
            }
          />
        </View>
      </View>
      <View
        onLayout={[Function]}
      />
    </View>
    <View
      collapsable={false}
      forwardedRef={[Function]}
      handlerTag={2}
      handlerType="TapGestureHandler"
      onGestureHandlerEvent={[Function]}
      onGestureHandlerStateChange={[Function]}
      pointerEvents="auto"
      style={
        {
          "transform": [
            {
              "translateX": -0,
            },
          ],
        }
      }
    >
      <View
        style={
          {
            "flexDirection": "row",
          }
        }
      >
        <View
          style={
            {
              "flexGrow": 1,
            }
          }
        >
          <Text
            onPress={[Function]}
            style={
              {
                "color": "#7f7f7f",
                "fontSize": 35,
              }
            }
          >
            12:00
          </Text>
          <Text
            style={
              {
                "color": "#a4a4a4",
              }
            }
          >
            testing
          </Text>
          <RNDateTimePicker
            date={1687817628067}
            displayIOS="default"
            enabled={true}
            is24Hour={true}
            mode="time"
            onChange={[Function]}
            onPickerDismiss={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
          />
        </View>
        <View
          style={
            {
              "alignItems": "center",
              "flexDirection": "row-reverse",
              "padding": 10,
            }
          }
        >
          <RCTSwitch
            accessibilityRole="switch"
            onChange={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "height": 31,
                "width": 51,
              }
            }
            testID="switch"
            value={true}
          />
        </View>
      </View>
    </View>
  </View>
</View>
`;
